apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs

```diff
@@ -0,0 +1,98 @@
+
+        partial void Modify_CreateSessionRequest(ref CreateSessionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_BatchCreateSessionsRequest(ref BatchCreateSessionsRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_GetSessionRequest(ref GetSessionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_ListSessionsRequest(ref ListSessionsRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_DeleteSessionRequest(ref DeleteSessionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_ExecuteSqlRequest(ref ExecuteSqlRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_ExecuteBatchDmlRequest(ref ExecuteBatchDmlRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_ReadRequest(ref ReadRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_BeginTransactionRequest(ref BeginTransactionRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_CommitRequest(ref CommitRequest request, ref CallSettings settings)=>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_RollbackRequest(ref RollbackRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_PartitionQueryRequest(ref PartitionQueryRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        partial void Modify_PartitionReadRequest(ref PartitionReadRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Session);
+
+        private static void ApplyResourcePrefixHeader(ref CallSettings settings, string resource)
+        {
+            // If we haven't been given a resource name, just leave the request as it is.
+            if (string.IsNullOrEmpty(resource))
+            {
+                return;
+            }
+
+            // Try to parse the resource as a database name.
+            DatabaseName database;
+            if (!DatabaseName.TryParse(resource, out database))
+            {
+                // If that fails, the resource name is likely a session name.
```

Comment:
Do we definitely want to use Parse rather than TryParse here? If we have a call with a resource which is neither a session or a database name, do we definitely want it to throw here?

Moreover, might it not make sense to use different methods here, one expecting a session name and one expecting a database name? I'd expect us to know which should be used on a per-request basis.

Reply:
I guess you're right and it's overkill to fail the request here if an appropriate resource name is not specified. I guess that's not the job of the client library.

I split the function into two as you suggested and now ignoring the failure to parse.